### ISSUE TITLE
Offer `pytensor.tensor` functionality at the pytensor root level

### ISSUE THREAD

**Issue by ricardoV94** [pymc-devs/pytensor#635]
### Description

Make import `pytensor as pt` basically work like todays `import pytensor.tensor as pt`. 

For users that's all they need, besides functionality like `scan`, `ifelse`, `function`, `OpFromGraph`, `shared` that should also be available at the root level (see #332)

Internally things can still be implemented inside the `tensor` submodule.

Does anyone have strong objections or see why this may not be possible?

**Comment by maresb** [comment]
I'm weakly :-1:.

I don't see any technical obstacles. But is there any benefit beyond saving users a few keystrokes?

This would introduce yet another layer of indirection. It's already pretty tricky to find the definitions of stuff. Like to find the definition of `pt.dscalar` I have to To look in `__init__.py`, and then I search for `dscalar` and there's no result, because it's actually pulled in implicitly `pytensor.tensor.type import *`. There are lots of other `import *` statements, so the only way I could find out which one was with a raw text search.

With various linters and type checkers I've gotten into the habit of always importing things directly from their definition, and so my personal feeling is that these types of shortcuts cost me way more time than they save.

TLDR: keystrokes are cheap, directness and clarity are priceless

Counterproposal: eliminate `*`-imports from the codebase.

**Comment by ricardoV94** [comment]
Writing `pytensor.tensor` is a PITA, only surpassed by `matplotlib.pyplot`.

Importing from source is fine for dev work but not user work (like Jupyter notebook workflows).

Why should the main functionality of a package require a double level import?

**Comment by ricardoV94** [comment]
The counter proposal doesn't make sense to me. 

Are you suggesting one to do `from pytensor.tensor.math/basic/elemwise import add, dot, cos, zeros`...?

**Comment by maresb** [comment]
> Importing from source is fine for dev work but not user work, such as Jupyter notebook workflows.

Ya, that's actually a really good argument.

**Comment by twiecki** [comment]
I'm strongly :+1:.

**Comment by jessegrabowski** [comment]
I like both the proposal and the counter-proposal.

While I don't love it, I'll also throw into the mix the `import statsmodels.api as sm` pattern. They provide the `api` module as a way to organize and present all the user-facing stuff in a clean way, while still letting them organize the rest of the package in a dev friendly way.

**Comment by ricardoV94** [comment]
@jessegrabowski can you expand on what the `api` module does? I am not familiar with it

**Comment by twiecki** [comment]
FWIW I always found `statsmodels`' import structure pretty odd and besides matplotlib I don't know another library that does it that way.

**Comment by maresb** [comment]
Just to brainstorm likely-bad ideas, I wonder if it would make sense to flip the `statsmodels` pattern upside down. Make the top-level `pytensor` namespace the user API. Then in order to ensure separation between the user API and the development library, we could move the actual code into `pytensor.lib`. That way we ensure no name conflicts, and the user can tab-complete all user API names.

**Comment by ricardoV94** [comment]
@Dhruvanshu-Joshi we should probably hold on until the next major release, perhaps when we drop the C code or make numba the default we can go ahead and also do this major change

**Comment by ricardoV94** [comment]
The introduction of `pytensor.xtensor` is an argument against this change :D

**Comment by jessegrabowski** [comment]
> Counterproposal: eliminate *-imports from the codebase.

Let's do this instead

**Comment by ricardoV94** [comment]
I'm against this due to `pytensor.xtensor` which is semantically pararellel to `pytensor.tensor`. We are no longer pushing a pure-numpy-like API in that sense, so it shouldn't be the default at the root scope. I'm closing this issue since I was the one proposing it, but if someone wants to champion it, feel free to reopen
